File: NetFramework/FlexPivot/CS/CustomizePage/UndoStack.cs
using System;
using System.Windows.Forms;
using System.Collections.Generic;

namespace CustomizePage
{
	public abstract class UndoStack
	{
		// ** fields

		List<object> _stack = new List<object>();
        int _current;
        bool _updating;

		// ** constants

		private const int MAX_STACK_SIZE = 12;

		// ** ctor

		public UndoStack()
		{
		}

        // ** abstracts

        public abstract object GetCurrentState();
        public abstract void ApplySavedState(object state);

		// ** object model

		public void Clear()
		{
			_stack.Clear();
			_current = 0;
		}
		public bool CanUndo
		{
			get { return _current > 0; }
		}
		public bool Undo()
		{
			if (!CanUndo) return false;
            if (_current == _stack.Count)
			{
				int ptr = _current;
				SaveState();
				_current = ptr;
			}
			_current--;
            _updating = true;
            RestoreState(_current);
            _updating = false;
			return true;
		}
		public bool CanRedo
		{
			get { return _current < _stack.Count - 1; }
		}
		public bool Redo()
		{
			if (!CanRedo) return false;
            _updating = true;
            _current++;
			RestoreState(_current);
            _updating = false;
            return true;
		}
        public bool SaveState()
        {
            if (_updating)
            {
                return false;
            }

            while (_stack.Count > _current)
            {
                _stack.RemoveAt(_stack.Count - 1);
            }
            while (_stack.Count >= MAX_STACK_SIZE)
            {
                _stack.RemoveAt(0);
            }
            object state = GetCurrentState();
            _stack.Add(state);
            _current = _stack.Count;
            return true;
        }
        bool RestoreState(int index)
        {
            if (index < 0 || index >= _stack.Count) return false;
            object state = _stack[index];
            ApplySavedState(state);
            return true;
        }
    }
}

File: NetFramework/FlexPivot/VB/CustomizePage/UndoStack.vb
Namespace CustomizePage
	Public MustInherit Class UndoStack
		' ** fields

		Private _stack As New List(Of Object)()
		Private _current As Integer
		Private _updating As Boolean

		' ** constants

		Private Const MAX_STACK_SIZE As Integer = 12

		' ** ctor

		Public Sub New()
		End Sub

		' ** abstracts

		Public MustOverride Function GetCurrentState() As Object
		Public MustOverride Sub ApplySavedState(ByVal state As Object)

		' ** object model

		Public Sub Clear()
			_stack.Clear()
			_current = 0
		End Sub
		Public ReadOnly Property CanUndo() As Boolean
			Get
				Return _current > 0
			End Get
		End Property
		Public Function Undo() As Boolean
			If Not CanUndo Then
				Return False
			End If
			If _current = _stack.Count Then
				Dim ptr As Integer = _current
				SaveState()
				_current = ptr
			End If
			_current -= 1
			_updating = True
			RestoreState(_current)
			_updating = False
			Return True
		End Function
		Public ReadOnly Property CanRedo() As Boolean
			Get
				Return _current < _stack.Count - 1
			End Get
		End Property
		Public Function Redo() As Boolean
			If Not CanRedo Then
				Return False
			End If
			_updating = True
			_current += 1
			RestoreState(_current)
			_updating = False
			Return True
		End Function
		Public Function SaveState() As Boolean
			If _updating Then
				Return False
			End If

			Do While _stack.Count > _current
				_stack.RemoveAt(_stack.Count - 1)
			Loop
			Do While _stack.Count >= MAX_STACK_SIZE
				_stack.RemoveAt(0)
			Loop
			Dim state As Object = GetCurrentState()
			_stack.Add(state)
			_current = _stack.Count
			Return True
		End Function
		Private Function RestoreState(ByVal index As Integer) As Boolean
			If index < 0 OrElse index >= _stack.Count Then
				Return False
			End If
			Dim state As Object = _stack(index)
			ApplySavedState(state)
			Return True
		End Function
	End Class
End Namespace
